### ISSUE TITLE
DecimalBinaryOp_03 fails on Linux in CI on `master` branch

### ISSUE THREAD

**Issue by jcouv** [dotnet/roslyn#34198]
I'm getting an error on Linux branch with `DecimalBinaryOp_03` in two of my PRs, and I see that other PRs are affected too.
It's effectively doing: `System.Console.WriteLine(12345678900000000M % 0.0000000012345678M);`
We expect `0.000000000983` but got `0.0000000009832122`.

Why did this appear in unrelated PRs (seems like CI should have prevented whatever caused this change from getting merged, unless we're not building in a controlled/isolated environment), and why only on the Linux legs?

Relates to https://github.com/dotnet/roslyn/issues/32576

 

 

```

Roslyn.Test.Utilities.ExecutionException : \nExecution failed for assembly '/opt/code/artifacts/tmp/Debug/RoslynTests'.\n

Expected: \n1007\n993\n7000\n142.85714285714285714285714286\n6\n-993\n-1007\n-7000\n-142.85714285714285714285714286\n-6\n123.0012300\n122.9987700\n0.15129000000000\n100000\n0.0000000\n12345678900000000.000000001235\n12345678899999999.999999998765\n15241577.6390794200000000\n10000000729000059778004901.796\n0.000000000983\n246913578.1246913578\n-0.1000000000\n15241578765584515.651425087878\n0.9999999991899999933660999449\n123456789.0123456789\n\n

Actual: 1007\n993\n7000\n142.85714285714285714285714286\n6\n-993\n-1007\n-7000\n-142.85714285714285714285714286\n-6\n123.0012300\n122.9987700\n0.15129000000000\n100000\n0.0000000\n12345678900000000.000000001235\n12345678899999999.999999998765\n15241577.6390794200000000\n10000000729000059778004901.796\n0.0000000009832122\n246913578.1246913578\n-0.1000000000\n15241578765584515.651425087878\n0.9999999991899999933660999449\n123456789.0123456789\n\n

 

Stack trace

  at Roslyn.Test.Utilities.Desktop.DesktopRuntimeEnvironment.Execute (System.String moduleName, System.String[] args, System.String expectedOutput) [0x00079] in <82d3df1bc9244dfab11b08c80afc1923>:0

  at Microsoft.CodeAnalysis.Test.Utilities.CompilationVerifier.Emit (System.String expectedOutput, System.Nullable`1[T] expectedReturnCode, System.String[] args, System.Collections.Generic.IEnumerable`1[T] manifestResources, Microsoft.CodeAnalysis.Emit.EmitOptions emitOptions, Microsoft.CodeAnalysis.Test.Utilities.Verification peVerify, Microsoft.CodeAnalysis.Test.Utilities.SignatureDescription[] expectedSignatures) [0x00055] in <82d3df1bc9244dfab11b08c80afc1923>:0

  at Microsoft.CodeAnalysis.Test.Utilities.CommonTestBase.Emit (Microsoft.CodeAnalysis.Compilation compilation, System.Collections.Generic.IEnumerable`1[T] dependencies, System.Collections.Generic.IEnumerable`1[T] manifestResources, Microsoft.CodeAnalysis.Test.Utilities.SignatureDescription[] expectedSignatures, System.String expectedOutput, System.Nullable`1[T] expectedReturnCode, System.String[] args, System.Action`1[T] assemblyValidator, System.Action`1[T] symbolValidator, Microsoft.CodeAnalysis.Emit.EmitOptions emitOptions, Microsoft.CodeAnalysis.Test.Utilities.Verification verify) [0x00016] in <82d3df1bc9244dfab11b08c80afc1923>:0

  at Microsoft.CodeAnalysis.Test.Utilities.CommonTestBase.CompileAndVerifyCommon (Microsoft.CodeAnalysis.Compilation compilation, System.Collections.Generic.IEnumerable`1[T] manifestResources, System.Collections.Generic.IEnumerable`1[T] dependencies, System.Action`1[T] sourceSymbolValidator, System.Action`1[T] assemblyValidator, System.Action`1[T] symbolValidator, Microsoft.CodeAnalysis.Test.Utilities.SignatureDescription[] expectedSignatures, System.String expectedOutput, System.Nullable`1[T] expectedReturnCode, System.String[] args, Microsoft.CodeAnalysis.Emit.EmitOptions emitOptions, Microsoft.CodeAnalysis.Test.Utilities.Verification verify) [0x00074] in <82d3df1bc9244dfab11b08c80afc1923>:0

  at Microsoft.CodeAnalysis.CSharp.Test.Utilities.CSharpTestBase.CompileAndVerify (Microsoft.CodeAnalysis.Compilation compilation, System.Collections.Generic.IEnumerable`1[T] manifestResources, System.Collections.Generic.IEnumerable`1[T] dependencies, System.Action`1[T] sourceSymbolValidator, System.Action`1[T] validator, System.Action`1[T] symbolValidator, Microsoft.CodeAnalysis.Test.Utilities.SignatureDescription[] expectedSignatures, System.String expectedOutput, System.Nullable`1[T] expectedReturnCode, System.String[] args, Microsoft.CodeAnalysis.Emit.EmitOptions emitOptions, Microsoft.CodeAnalysis.Test.Utilities.Verification verify) [0x00016] in <410994acfd424b9db2d2c066e1ae29e9>:0

  at Microsoft.CodeAnalysis.CSharp.Test.Utilities.CSharpTestBase.CompileAndVerify (Microsoft.CodeAnalysis.CSharp.Test.Utilities.CSharpTestSource source, System.Collections.Generic.IEnumerable`1[T] references, System.Collections.Generic.IEnumerable`1[T] manifestResources, System.Collections.Generic.IEnumerable`1[T] dependencies, System.Action`1[T] sourceSymbolValidator, System.Action`1[T] assemblyValidator, System.Action`1[T] symbolValidator, Microsoft.CodeAnalysis.Test.Utilities.SignatureDescription[] expectedSignatures, System.String expectedOutput, System.Nullable`1[T] expectedReturnCode, System.String[] args, Microsoft.CodeAnalysis.CSharp.CSharpCompilationOptions options, Microsoft.CodeAnalysis.CSharp.CSharpParseOptions parseOptions, Microsoft.CodeAnalysis.Emit.EmitOptions emitOptions, Roslyn.Test.Utilities.TargetFramework targetFramework, Microsoft.CodeAnalysis.Test.Utilities.Verification verify) [0x00034] in <410994acfd424b9db2d2c066e1ae29e9>:0

  at Microsoft.CodeAnalysis.CSharp.UnitTests.CodeGen.CodeGenTests.DecimalBinaryOp_03 () [0x0000e] in <713891092dbf4c60917eca0a107b2bf5>:0

  at (wrapper managed-to-native) System.Reflection.RuntimeMethodInfo.InternalInvoke(System.Reflection.RuntimeMethodInfo,object,object[],System.Exception&)

  at System.Reflection.Run

````

**Comment by tannergooding** [comment]
@jcouv, were you able to determine why these just started failing or why they only started failing for Linux?

The fixes to `System.Decimal` were made in managed code and should be deterministic across all platforms/architectures.

**Comment by jcouv** [comment]
Not yet. I had other plans for the weekend and wanted to check with folks first in case there's some important clue.

Just to clarify: is the change in behavior expected (ie. the new result is more correct)? In that case, the question may be why isn't the change affecting Core/windows CI legs too.

**Comment by jcouv** [comment]
Looking at https://github.com/dotnet/coreclr/issues/22046 (combo issue Aleksey filed, with 5 scenarios), it sounds like this behavior (`DecimalBinaryOp_03`) changed some time between Core 2 and 3, and is not affected by the recent fix (https://github.com/dotnet/coreclr/pull/22070).

@tannergooding Do you know what PR caused that change (increased precision)? 
I see that `Decimal.Remainder` was reimplemented, but that was back in Oct 2018 ([New Decimal.Remainder implementation](https://github.com/dotnet/coreclr/pull/20305)). Would it be possible to run `System.Console.WriteLine(12345678900000000M % 0.0000000012345678M);` before and after that change to confirm that is the root?

**Comment by jcouv** [comment]
Assigned to Jared following our discussion.
Tagging @agocke

**Comment by jaredpar** [comment]
This should definitely not be showing up in our tests at the moment. In order for the `decimal` precision change to be observed in our tests the following things must **all** happen:

1. 3.0 DotNet runtime installed on the machine.
1. 3.0 DotNet runtime installed in the same location as the 2.1 `dotnet` used to run CI. Multi-level lookup is explicitly disabled in our CI hence a machine wide install is not enough here.
1. The environment must be configured to allow for roll forward. This is disabled by default and xunit has a target framework of netcoreapp2.0 hence it can't silently run on netcoreapp3.0.

This is doubly hard for us to accidentally observe because our Linux tests run in a docker container. Basically this is quite puzzling. 

I think the most likely cause here is docker container pollution from the DIM branch. It seems plausible that the docker container name, and all other inputs, are the same between the master and DIM branch. That would mean for a given Linux VM whichever branch ran first established the dotnet setup in the enlistment. Assuming DIM went first that would mean both 2.1 and 3.0 were installed side by side in the `.dotnet` directory. The next run from master which inherited this setup would have both runtimes, roll forward enabled and trip this bug. 

Think the solution here is to change the docker setup so that it takes the branch name as an input. That would give us the proper isolation.

**Comment by agocke** [comment]
@jaredpar This seems highly likely to be correct to me. I will also test this locally.

**Comment by jcouv** [comment]
What's the benefit of enabling roll forward? I would have assumed we preferred a more controlled environment (ie. specific versions).

**Comment by jaredpar** [comment]
@jcouv 

> What's the benefit of enabling roll forward? 

It was a quick fix to unblock DIM work while I got the correct solution implemented in master. It's properly marked with a `PROTOTYPE` comment and not meant to ever go back to `master`. 

> I would have assumed we preferred a more controlled environment (ie. specific versions).

Mostly. There is some amount of roll forward in our repository as it's the only rational way to run certain operations. For example we must roll xunit forward to 3.0 when running DIM tests. But we try to be fairly specific about it vs. setting the global roll forward environment variable.

**Comment by chsienki** [comment]
Hmm. Thinking this through, I'm not sure if docker is actually the problem here.

Ultimately we share a host directory into the container, we don't git clone inside of the container. That means we could share the filesystem between runs, even though we're creating a brand new container. I'm not sure if VSTS inherently allows sharing, but it reused the file system between container runs it would explain this.

Let me do some digging.

**Comment by chsienki** [comment]
Did some digging. I hadn't realized that the failures were occurring in Linux - **Mono!** 

That means they're not directly related to @tannergooding 's changes in coreclr, or the DIM work leaking into other builds. Turns out Mono ported some of those changes over, which is why we're seeing the failures on nightly there. (See https://github.com/mono/mono/commit/e42bf95d14b7ea16abcdff99d769e9f250ec2fd0#diff-0271e682606c3e81e9ca8944c00871a9) That means we can just skip the test for Mono or conditionally change the expected value depending on the runtime.

Thankfully this means the docker / build infra still seems sane, which is always a pleasant surprise.

**Comment by jcouv** [comment]
Thanks Chris for catching this. It looks like I've shared bad information. Went back to look at the problematic PRs and the problem was indeed was Linux/Mono (not Linux/Core as I thought). Sorry about that.

![image](https://user-images.githubusercontent.com/12466233/54644084-b7553100-4a55-11e9-8a3e-5baf5b02301c.png)

**Comment by jcouv** [comment]
Closing as duplicate of https://github.com/dotnet/roslyn/issues/32576 (A number of C# and VB compiler unit-tests utilizing decimal numbers fail due to a baseline difference when run against netcoreapp3.0)
